src/MotionControl.c

```diff
@@ -755,88 +796,236 @@ void Ros_MotionControl_IncMoveLoopStart() //<-- IP_CLK priority task
                         }
 
-                        g_Ros_Controller.ctrlGroups[i]->q_time = inc_data_time;
+                        // Unlock the q
+                        mpSemGive(q->q_lock);
                     }
                     else
                     {
-                        moveData.grp_pos_info[i].pos_tag.data[2] = 0;
-                        moveData.grp_pos_info[i].pos_tag.data[3] = MP_INC_PULSE_DTYPE;
-                        moveData.grp_pos_info[i].pos_tag.data[4] = 0;
+                        Ros_Debug_BroadcastMsg("ERROR: Can't get data from queue. Queue is locked up (Group #%d)", g_Ros_Controller.ctrlGroups[i]->groupNo);
                         bzero(&moveData.grp_pos_info[i].pos, sizeof(LONG) * MP_GRP_AXES_NUM);
+                        continue;
+                    }
+                }
+            }
+
+            for (i = 0; i < g_Ros_Controller.numGroup; i++)
+            {
+                memcpy(newPulseInc[i], moveData.grp_pos_info[i].pos, sizeof(LONG) * MP_GRP_AXES_NUM);
+
+                // record the speed associate with the next amount of pulses
+                if (queueRead[i])
+                {
+                    for (axis = 0; axis < MP_GRP_AXES_NUM; axis++)
+                    {
+                        maxSpeed[i][axis] = abs(newPulseInc[i][axis]);
+                        maxSpeedRemain[i][axis] = abs(newPulseInc[i][axis]);
+                    }
+                }
+
+                // Check if pulses are missing from last increment.
+                // Get the current controller command position and substract the previous command position
+                // and check if it matches the amount if increment sent last cycle.  If it doesn't then
+                // some pulses are missing and the amount of unprocessed pulses needs to be added to this cycle.
+                mpGetPulsePos(&ctrlGrpData, &pulsePosData);
+                isMissingPulse = FALSE;
+                for (axis = 0; axis < MP_GRP_AXES_NUM; axis++)
+                {
+                    // Check how many pulses we processed from last increment
+                    processedPulses[axis] = pulsePosData.lPos[axis] - prevPulsePosData[i].lPos[axis];
+                    prevPulsePosData[i].lPos[axis] = pulsePosData.lPos[axis];
+
+                    // Remove those pulses from the amount to process.  
+                    // If everything was processed, then there should by 0 pulses left. Otherwise FSU Speed limit prevented processing
+                    toProcessPulses[i][axis] -= processedPulses[axis];
+                    if (toProcessPulses[i][axis] != 0)
+                        isMissingPulse = TRUE;
+
+                    // Add the new pulses to be processed for this iteration 
+                    toProcessPulses[i][axis] += newPulseInc[i][axis];
+
+                    if (toProcessPulses[i][axis] != 0)
+                        hasUnprocessedData = TRUE;
+                }
```

Comment:
@EricMarcil: could we assume the ratio between submitted and processed pulses could be used to calculate the amount / percentage by which the FSU is currently throttling robot performance?

Reply:
@gavanderhoorn It might give some idea but it is difficult to evaluate.  The FSU Speed Limit is usually a restriction in mm/s in the Cartesian space but the MotoRos is working in the joint space.  It would probably be easier to just observe the motion resulting speed at the end of arm.